### DISCUSSION TITLE
GitHub seamlessly enhances IT developers’ coding experience!

### DISCUSSION THREAD

**Discussion by nikhil-thampi** [community/community#28796] [Discover: GitHub Best Practices]
There is often an expectation that every junior developer, by default, knows about source code management and security analysis of source code at the beginning of their career. But in reality, many developers come from different backgrounds. For instance, I received my campus placement after completing my engineering degree in Electronics and Communication, with basic C++ knowledge. In my university days, I had friends with Computer Science and IT Engineering degrees, but a sizable portion of them had diverse backgrounds, like Mechanical, Civil, Electrical, and even Chemical Engineering, with basic knowledge of popular computer languages, along with a spirit to succeed. We were lucky to have amazing mentors during the early part of our careers, who introduced the difference between coding in isolation vs coding for an enterprise, and I am sure this is the story of millions of developers across the world!

Thinking back to the early days in my career, I wanted to share a few thoughts and questions I had, specifically in the areas where GitHub really shines.

### 1.) Why do I have to push code to GitHub every time, instead of updating the code in production? 

![Production](https://user-images.githubusercontent.com/49915749/183471490-fb29c837-3c2e-4c96-b34d-5725b0505d92.png)

The above question is a classic one that many new developers have on their mind while working in an enterprise. Let's imagine a fictional scenario, where a developer has created an application for an insurance company. Unfortunately, there is a bug in the application and it has resulted in the calculation of a higher premium for automobile insurance. We all are human and errors happen. But the impact of one person’s error can adversely affect millions of families, unless fixed. This is where source code management systems like GitHub come into picture, and can be the greatest resource for developers. GitHub helps identify the changes in each version of your code and passes these versions to others who may be working on said code (ie: Testing teams, DevOps Engineers, Security Engineers) via different stages (like functional tests, performance tests, security tests) before sending them to the production environment. This also ensures that even in the worst case scenarios where you are not available, other members of your team can pitch in and make sure that customers will not pay additional money due to a simple calculation mistake during the code creation!

### 2.) I am fine with the testing team verifying my code but why is everyone talking about DevSecOps?

![DevSecOps](https://user-images.githubusercontent.com/49915749/183473011-97f699f5-d547-49ef-82f2-60a528bee53e.png)

Now that you understand the value of GitHub in managing your source code, let's talk about some of the other people you might interact with on a regular basis during your code promotion cycle. Two of the most common are Developers and Testers. However, looking at an even larger landscape of the code cycle, we can consider a couple of other personas you might regularly interact with:

* DevOps Engineers 
* Security Engineers

Now the big question is why do we need them? Before discussing the term DevSecOps, let's take a look at each of these roles individually.

**DevOps Engineers** - You might be more than happy to move your code from a developer environment to production yourself. But in practice, DevOps Engineers ensure that enough checks and balances are in place while moving code to production. They typically do this at scale and in an automated fashion. In areas of critical importance like financial services, telecom, government etc, it is a routine affair to have a third party team handle all the operational aspects with proper governance in place. DevOps Engineering ensures operational stability is maintained during code promotion, by using automated tools like [GitHub Actions](https://github.com/features/actions).

**Security Engineers** - They make sure that your code is not vulnerable to security breaches. Their main objective is to make sure that the enterprise infrastructure is not compromised due to code vulnerability.  

Traditionally the above two roles work in isolation without much collaboration with each other, which has resulted in friction between the teams and loss of productivity in the enterprises. So the world has moved to a more collaborative method of working between the developers; DevOps engineers and Security engineers, making the term [DevSecOps](https://resources.github.com/devsecops/fundamentals/guide/) popular among IT circles. 

### 3.) If the above concepts and roles are there to help me, why is everyone still asking to fix the code ASAP?

![SDLC](https://user-images.githubusercontent.com/49915749/183472315-85042eea-1d5a-4c6c-9d26-3780f1e2aa03.png)

If developers fix the code themselves during the development stage, then overhead costs can be mitigated. Let's assume that you are a developer working in banking software. In your code, you accidentally introduce a bug and during the verification stage, you find and rectify it. You may have done this within minutes or hours.

As shown above, when your bug is moved further towards the right, it requires time and effort from more people to identify and fix said bug. More people working on identifying an issue, means additional accrued costs for things like salary, test execution time in various code promotion stages, associated server running costs, code movement time, etc. Even with all these verification processes, some bugs may still get into production. This may result in breaches and potentially millions of dollars getting lost as part of the breach. This is why enterprises prefer to identify problems early on: dealing with the issue swiftly and quickly, and keeping costs at a minimum. 

### 4.) I understand the need for different checkpoints and the need to fix issues early, but it's affecting my productivity!

This is a common and logical thought; the more people who are involved, the longer it may take. But fear not, let's talk about the productivity part and how GitHub can help you. Yes, in normal cases, productivity may take a hit if the above personas work in isolation.  

However, GitHub enables teams to break silos and work collaboratively. It also ensures that security issues are addressed at an earlier stage of the code development cycle, so that the cost of fixing those issues will be automatically reduced. The security testing and remediation steps of GitHub happens in the same workflow that the developer uses to build the code. This means that instead of a productivity roadblock, GitHub acts as a booster to increase productivity and helps developers fix security-related issues during earlier stages of the software development life cycle.

Now that we have solved some of the basic queries which may arise in a new developer’s mind, we would love to hear your thoughts and experiences related to DevSecOps and GitHub. In the next episode of this blog series, we will discover more about GitHub Advanced Security(GHAS) and various elements of GHAS namely code scanning, secret scanning, Dependabot, and Security Overview.

**Comment by ritayong99** [comment]
thank you for sharing your knowledge that's gonne be so helpful 
 [Appvalley](https://getappvalley.com/) [TutuApp](https://tutuappx.com/) [Tweakbox](https://tweakbox.mobi/)

**Reply by BlueDevils0** [reply]
Sees good let see what happens.

**Comment by Emanuelcarvalho2004** [comment]


**Reply by poposet** [reply]
fvh;av\77 uao44fud]

 9u9w

**Reply by poposet** [reply]
tttttttttttttttttttttttttttttttttttttttttttt

**Comment by rdotterer09** [comment]
I asked for my contract to be upgraded so I would have access, what I received is a node ui something, im not a coder, but I understand enough to know I received an upgraded EVM , what can I do? They paid themselves half million from the contract and I still don't have access, it's on uniswap, Binance, aave, and many more, I need some serious help

**Comment by rdotterer09** [comment]
I understand the need for all that are involved in the maintenance of the contract, what nobody is understanding is I want 1/4 of all assets within this contract put in my bank account then the contract can continue on as usual, this is my retirement, I'm ready to retire, have I now cleared up any misunderstanding? I know all of you that are involved in this contract are doing so to make a living I have no intentions of taking that away, it benefits me also to keep this contract moving forward, what really makes me angry is being locked out of the opportunity to gain access to my own money as I'm sure all of you will agree, when someone else has control of your funds and tells you repeatedly NO you can't have any, in my head I'm asking, why the F not IT'S MINE! and I'm ready to retire NOW, so is there a way to accomplish what I ask for? Please consider what I have said and let me know where we will go from here

Thank You,
Rhonda Dotterer

**Comment by poposet** [comment]
/rg[
up
nnu,
2/e=yp4\= e
omu[npubu9[g\5f\]=/+3

**Comment by liyj0** [comment]
感谢您的来信，我会尽快查看的。

**Reply by Julio041891** [reply]
5056160009663539

**Comment by Kashumy** [comment]
when I made my own android app. I was surprised to see an inscription about the security status of the application.
If anyone knows what's going on, let them dowload my application and will decompile it using apktool m. 
If enyone have mobile device. 

> https://kashumy.github.io/My_aplications/Index.html

**Comment by jonyboy11** [comment]
Great reflections! What you wrote about the journey from being a beginner developer to understanding GitHub, DevOps, and DevSecOps really resonates. In many ways, it reminds me of the difference between gross and net salary: at first, you only see the surface (the gross number, or in coding, just “my app runs”), but as you gain experience, you discover all the hidden layers, taxes, contributions, security, governance — that determine what actually remains in your pocket, or in production.

That’s exactly why salary calculator for Italy offered by [calcolo-stipendionetto.it ](https://calcolo-stipendionetto.it/)are helpful for workers, and GitHub is equally essential for developers: both give you clarity and prevent costly surprises later on, whether in your paycheck or in your code lifecycle.

**Comment by subupatayach** [comment]
gmail, and all mail concert my mail id account acctiovets online support
and all world all contrary all bank accounts actiovtes on online 0balance
acctiovets and online international transaction money created my mail 🆔
accounts actiovtes and segunail tower setelight everything online include
support my mail 🆔 accounts on online.

**Comment by VictorMoni** [comment]
Nice reflection! Good topic to read! 🚀🔟⭐️
